crates/ruma-common/src/events/room/message/sanitize.rs

```diff
@@ -0,0 +1,202 @@
+/// Sanitize the given HTML string.
+///
+/// This removes the [tags and attributes] that are not listed in the Matrix specification.
+///
+/// It can also optionally remove the [rich reply fallback].
+///
+/// [tags and attributes]: https://spec.matrix.org/v1.2/client-server-api/#mroommessage-msgtypes
+/// [rich reply fallback]: https://spec.matrix.org/v1.2/client-server-api/#fallbacks-for-rich-replies
+#[cfg(feature = "sanitize")]
+pub fn sanitize_html(s: &str, remove_reply_fallback: RemoveReplyFallback) -> String {
+    let allowed_tags = [
+        "font",
+        "del",
+        "h1",
+        "h2",
+        "h3",
+        "h4",
+        "h5",
+        "h6",
+        "blockquote",
+        "p",
+        "a",
+        "ul",
+        "ol",
+        "sup",
+        "sub",
+        "li",
+        "b",
+        "i",
+        "u",
+        "strong",
+        "em",
+        "strike",
+        "code",
+        "hr",
+        "br",
+        "div",
+        "table",
+        "thead",
+        "tbody",
+        "tr",
+        "th",
+        "td",
+        "caption",
+        "pre",
+        "span",
+        "img",
+        "details",
+        "summary",
+    ];
+
+    let allowed_attributes = [
+        ("font", vec!["data-mx-bg-color", "data-mx-color", "color"]),
+        ("span", vec!["data-mx-bg-color", "data-mx-color", "data-mx-spoiler"]),
+        ("a", vec!["name", "target", "href"]),
+        ("img", vec!["width", "height", "alt", "title", "src"]),
+        ("ol", vec!["start"]),
+        ("code", vec!["class"]),
+    ];
+
+    let allowed_urls = ["http", "https", "ftp", "mailto", "magnet"];
```

Comment:
```suggestion
    let allowed_schemes = ["http", "https", "ftp", "mailto", "magnet", "matrix"];
```

(dont add to commit batch, the var name has to be renamed below)

Reply:
Also, this is likely to remove `element://` schemes, which i have seen mentioned a bunch of times.

See my comment above about how this might affect derivative clients.

Reply:
Hmm the `matrix` scheme not being present is actually an omission from the spec. We should definitely add it though…

Reply:
Interesting, then i'll create an spec issue for that.

Reply:
https://github.com/matrix-org/matrix-spec/issues/1108

Reply:
And actually `mxc` is missing too, because it applies to both links and images. That means we're not fully compliant with the spec, I should probably mention it in the docs.

By the way, I should probably implement [MSC2398](https://github.com/matrix-org/matrix-spec-proposals/pull/2398) while I'm at it

Reply:
> By the way, I should probably implement https://github.com/matrix-org/matrix-spec-proposals/pull/2398 while I'm at it

Behind a feature flag, and I recommend doing it in a different PR

Reply:
Or we could use something else than ammonia, and be fully compliant with the spec.

Reply:
I made https://github.com/ruma/ruma/issues/1165, but you also mentioned alternative sanitization crates.

Reply:
The ones I mentioned are [html_sanitizer](https://crates.io/crates/html_sanitizer) and [sanitize_html](https://crates.io/crates/sanitize_html).

They are also whitelist-based but allow for configuring the behavior per tag instead of globally.

It should be noted that right now only a fully custom sanitizer would allow to apply this rule from the spec:

> Tags must not be nested more than 100 levels deep.

Reply:
I implemented a custom sanitizer but I don't expose it as this PR is only interested in the API to sanitize messages according to the spec.

It could be moved, exported and improved with #1165 as part of a more complete feature.